src/mpid/ch4/netmod/ofi/ofi_recv.h

```diff
@@ -134,6 +139,22 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_irecv(void *buf,
     MPIR_Datatype_add_ref_if_not_builtin(datatype);
 
     recv_buf = (char *) buf + dt_true_lb;
+    MPL_gpu_query_pointer_attr(recv_buf, &attr);
+    if (data_sz && attr.type == MPL_GPU_POINTER_DEV) {
+        if (!MPIDI_OFI_ENABLE_HMEM) {
+            /* Fallback path, create a CONTIGUOUS host buf to receive data.
+             * Move data from host buf to gpu buf when receive completes. */
+            void *host_buf = NULL;
+            MPL_gpu_malloc_host(&host_buf, data_sz);
```

Comment:
This is a very large memory allocation.  Shouldn't it be pipelined?

Reply:
But I think this path will be replaced in future by pipeline work. At this point, we don't have pipeline infrastructure, so here is to make sure GPU data can be transferred.

Reply:
No, it'll be replaced by the iQueue infrastructure that'll maintain a pool of buffers for you.  But it'll not add pipelining.  You'll need to do pipelining here.  What I was suggesting was that you can malloc/free every time, but you still need to maintain state about how many bytes have been transferred and how many are left.  Once iQueue is ready, you can replace that malloc/free with getting buffers from the iQueue pool.

Reply:
I will add `FIXME` for future work, large message should go through chunked pipeline and use fi_read.